README.md
[![Go Doc][godoc-image]][godoc-url]
[![Build Status][workflow-image]][workflow-url]
[![Go Report Card][goreport-image]][goreport-url]
[![Test Coverage][coverage-image]][coverage-url]
[![Maintainability][maintainability-image]][maintainability-url]

# observer

This package can be used for building observable applications in Go.
It aims to unify three pillars of observability in one single package that is _easy-to-use_ and _hard-to-misuse_.

This package leverages the [OpenTelemetry](https://opentelemetry.io) API.
OpenTelemetry is a great initiative that has brought all different standards and APIs for observability under one umbrella.
However, due to the requirements for interoperability with existing systems, OpenTelemetry is complex and hard to use by design!
Many packages, configurations, and options make the developer experience not so pleasant.
Furthermore, due to the changing nature of this project, OpenTelemetry specification changes often so does the Go library for OpenTelemetry.
In my humble opinion, this is not how a single unified observability API should be.
Hopefully, many of these issues will go away once the API reaches to v1.0.0.
This package intends to provide a very minimal and yet practical API for observability by hiding the complexity of configuring and using OpenTelemetry API.

An Observer encompasses a logger, a meter, and a tracer.
It offers a single unified developer experience for enabling observability.

## The Three Pillars of Observability

### Logging

Logs are used for _auditing_ purposes (sometimes for debugging with limited capabilities).
When looking at logs, you need to know what to look for ahead of the time (known unknowns vs. unknown unknowns).
Since log data can have any arbitrary shape and size, they cannot be used for real-time computational purposes.
Logs are hard to track across different and distributed processes. Logs are also very expensive at scale.

### Metrics

Metrics are _regular time-series_ data with _low and fixed cardinality_.
They are aggregated by time. Metrics are used for **real-time** monitoring purposes.
Using metrics with can implement **SLIs** (service-level indicators), **SLOs** (service-level objectives), and automated alerts.
Metrics are very good at taking the distribution of data into account.
Metrics cannot be used with _high-cardinality data_.

### Tracing

Traces are used for _debugging_ and _tracking_ requests across different processes and services.
They can be used for identifying performance bottlenecks.
Due to their very data-heavy nature, traces in real-world applications need to be _sampled_.
Insights extracted from traces cannot be aggregated since they are sampled.
In other words, information captured by one trace does not tell anything about how this trace is compared against other traces, and what is the distribution of data.

## Quick Start

For the examples below, you can use the following `docker-compose.yml` file to bring up an observability stack:

```bash
git clone https://github.com/moorara/docker-compose.git
cd docker-compose/observability
docker-compose up -d
```

<details>
  <summary>Example: Prometheus & Jaeger</summary>

```go
package main

import (
  "context"
  "net/http"
  "time"

  "github.com/moorara/observer"
  "go.opentelemetry.io/otel/api/baggage"
  "go.opentelemetry.io/otel/api/metric"
  "go.opentelemetry.io/otel/label"
  "go.uber.org/zap"
)

type instruments struct {
  reqCounter  metric.Int64Counter
  reqDuration metric.Float64ValueRecorder
}

func newInstruments(meter metric.Meter) *instruments {
  mm := metric.Must(meter)

  return &instruments{
    reqCounter:  mm.NewInt64Counter("requests_total", metric.WithDescription("the total number of requests")),
    reqDuration: mm.NewFloat64ValueRecorder("request_duration_seconds", metric.WithDescription("the duration of requests in seconds")),
  }
}

type server struct {
  observer    observer.Observer
  instruments *instruments
}

func (s *server) Handle(ctx context.Context) {
  // Tracing
  ctx, span := s.observer.Tracer().Start(ctx, "handle-request")
  defer span.End()

  start := time.Now()
  s.fetch(ctx)
  s.respond(ctx)
  duration := time.Now().Sub(start)

  labels := []label.KeyValue{
    label.String("method", "GET"),
    label.String("endpoint", "/user"),
    label.Uint("statusCode", 200),
  }

  // Metrics
  s.observer.Meter().RecordBatch(ctx, labels,
    s.instruments.reqCounter.Measurement(1),
    s.instruments.reqDuration.Measurement(duration.Seconds()),
  )

  // Logging
  s.observer.Logger().Info("request handled successfully.",
    zap.String("method", "GET"),
    zap.String("endpoint", "/user"),
    zap.Uint("statusCode", 200),
  )
}

func (s *server) fetch(ctx context.Context) {
  _, span := s.observer.Tracer().Start(ctx, "read-database")
  defer span.End()

  time.Sleep(50 * time.Millisecond)
}

func (s *server) respond(ctx context.Context) {
  _, span := s.observer.Tracer().Start(ctx, "send-response")
  defer span.End()

  time.Sleep(10 * time.Millisecond)
}

func main() {
  // Creating a new Observer and set it as the singleton
  obsv := observer.New(true,
    observer.WithMetadata("my-service", "0.1.0", "production", "ca-central-1", map[string]string{
      "domain": "auth",
    }),
    observer.WithLogger("info"),
    observer.WithPrometheus(),
    observer.WithJaeger("localhost:6831", "", "", ""),
  )
  defer obsv.End(context.Background())

  srv := &server{
    observer:    obsv,
    instruments: newInstruments(obsv.Meter()),
  }

  // Creating a context
  ctx := context.Background()
  ctx = baggage.NewContext(ctx,
    label.String("tenant", "1234"),
  )

  srv.Handle(ctx)

  // Serving metrics endpoint
  http.Handle("/metrics", obsv)
  http.ListenAndServe(":8080", nil)
}
```

Here are the logs from stdout :

```json
{"level":"info","timestamp":"2020-08-29T21:10:47.763781-04:00","caller":"example/main.go:57","message":"request handled successfully.","domain":"auth","environment":"production","logger":"my-service","region":"ca-central-1","version":"0.1.0","method":"GET","endpoint":"/user","statusCode":200}
```

And here are the metrics reported at http://localhost:8080/metrics :

```
# HELP request_duration_seconds the duration of requests in seconds
# TYPE request_duration_seconds histogram
request_duration_seconds_bucket{endpoint="/user",method="GET",statusCode="200",le="+Inf"} 1
request_duration_seconds_sum{endpoint="/user",method="GET",statusCode="200"} 0.065279047
request_duration_seconds_count{endpoint="/user",method="GET",statusCode="200"} 1
# HELP requests_total the total number of requests
# TYPE requests_total counter
requests_total{endpoint="/user",method="GET",statusCode="200"} 1
```

You can also verfiy a trace is reported to Jaeger by visiting http://localhost:16686 .
</details>

<details>
  <summary>Example: OpenTelemetry Collector</summary>

```go
package main

import (
  "context"
  "time"

  "github.com/moorara/observer"
  "go.opentelemetry.io/otel/api/baggage"
  "go.opentelemetry.io/otel/api/metric"
  "go.opentelemetry.io/otel/label"
  "go.uber.org/zap"
)

type instruments struct {
  reqCounter  metric.Int64Counter
  reqDuration metric.Float64ValueRecorder
}

func newInstruments(meter metric.Meter) *instruments {
  mm := metric.Must(meter)

  return &instruments{
    reqCounter:  mm.NewInt64Counter("requests_total", metric.WithDescription("the total number of requests")),
    reqDuration: mm.NewFloat64ValueRecorder("request_duration_seconds", metric.WithDescription("the duration of requests in seconds")),
  }
}

type server struct {
  observer    observer.Observer
  instruments *instruments
}

func (s *server) Handle(ctx context.Context) {
  // Tracing
  ctx, span := s.observer.Tracer().Start(ctx, "handle-request")
  defer span.End()

  start := time.Now()
  s.fetch(ctx)
  s.respond(ctx)
  duration := time.Now().Sub(start)

  labels := []label.KeyValue{
    label.String("method", "GET"),
    label.String("endpoint", "/user"),
    label.Uint("statusCode", 200),
  }

  // Metrics
  s.observer.Meter().RecordBatch(ctx, labels,
    s.instruments.reqCounter.Measurement(1),
    s.instruments.reqDuration.Measurement(duration.Seconds()),
  )

  // Logging
  s.observer.Logger().Info("request handled successfully.",
    zap.String("method", "GET"),
    zap.String("endpoint", "/user"),
    zap.Uint("statusCode", 200),
  )
}

func (s *server) fetch(ctx context.Context) {
  _, span := s.observer.Tracer().Start(ctx, "read-database")
  defer span.End()

  time.Sleep(50 * time.Millisecond)
}

func (s *server) respond(ctx context.Context) {
  _, span := s.observer.Tracer().Start(ctx, "send-response")
  defer span.End()

  time.Sleep(10 * time.Millisecond)
}

func main() {
  // Creating a new Observer and set it as the singleton
  obsv := observer.New(true,
    observer.WithMetadata("my-service", "0.1.0", "production", "ca-central-1", map[string]string{
      "domain": "auth",
    }),
    observer.WithLogger("info"),
    observer.WithOpenTelemetry("localhost:55680", nil),
  )
  defer obsv.End(context.Background())

  srv := &server{
    observer:    obsv,
    instruments: newInstruments(obsv.Meter()),
  }

  // Creating a context
  ctx := context.Background()
  ctx = baggage.NewContext(ctx,
    label.String("tenant", "1234"),
  )

  srv.Handle(ctx)

  // Wait before exiting
  fmt.Scanln()
}
```

Here are the logs from stdout :

```json
{"level":"info","timestamp":"2020-08-29T22:00:33.274878-04:00","caller":"example/main.go:57","message":"request handled successfully.","domain":"auth","environment":"production","logger":"my-service","region":"ca-central-1","version":"0.1.0","method":"GET","endpoint":"/user","statusCode":200}
```

You can verify metrics are reported to OpenTelemetry collector by visiting http://localhost:8889/metrics :

```
# HELP requests_total the total number of requests
# TYPE requests_total gauge
requests_total{endpoint="/user",method="GET",statusCode="200"} 1
```

You can also verfiy OpenTelemetry collector reported a trace to Jaeger by visiting http://localhost:16686 .
</details>

## Options

Most options can be set through environment variables.
This lets SRE people change how the observability pipeline is configured without making any code change.

Options set explicity in the code will override those set by environment variables.

| Environment Variable | Description |
|----------------------|-------------|
| `OBSERVER_NAME` | The name of service or application. |
| `OBSERVER_VERSION` | The version of service or application. |
| `OBSERVER_ENVIRONMENT` | The name of environment in which the service or application is running. |
| `OBSERVER_REGION` | The name of region in which the service or application is running. |
| `OBSERVER_TAG_*` | Each variable prefixed with `OBSERVER_TAG_` represents a tag for the service or application. |
| `OBSERVER_LOGGER_ENABLED` | Whether or not to create a logger (boolean). |
| `OBSERVER_LOGGER_LEVEL` | The verbosity level for the logger (`debug`, `info`, `warn`, `error`, or `none`). |
| `OBSERVER_PROMETHEUS_ENABLED` | Whether or not to configure and create a Prometheus meter (boolean). |
| `OBSERVER_JAEGER_ENABLED` | Whether or not to configure and create a Jaeger tracer (boolean). |
| `OBSERVER_JAEGER_AGENT_ENDPOINT` | The address to the Jaeger agent (i.e. `localhost:6831`). |
| `OBSERVER_JAEGER_COLLECTOR_ENDPOINT` | The full URL to the Jaeger HTTP Thrift collector (i.e. `http://localhost:14268/api/traces`). |
| `OBSERVER_JAEGER_COLLECTOR_USERNAME` | The username for Jaeger collector endpoint if basic auth is required. |
| `OBSERVER_JAEGER_COLLECTOR_PASSWORD` | The password for Jaeger collector endpoint if basic auth is required. |
| `OBSERVER_OPENTELEMETRY_ENABLED` | Whether or not to configure and create an OpenTelemetry Collector meter and tracer (boolean). |
| `OBSERVER_OPENTELEMETRY_COLLECTOR_ADDRESS` | The address to OpenTelemetry collector (i.e. `localhost:55680`). |

## OpenTelemetry

### Logging

_TBD_

### Metrics


CHANGELOG.md
# Changelog

## [v0.3.4](https://github.com/moorara/observer/tree/v0.3.4) (2020-09-25)

[Full Changelog](https://github.com/moorara/observer/compare/v0.3.3...v0.3.4)

**Merged pull requests:**

- Update http client example [\#70](https://github.com/moorara/observer/pull/70) ([moorara](https://github.com/moorara))
- Update OpenTelemetry modules to v0.12.0 [\#69](https://github.com/moorara/observer/pull/69) ([moorara](https://github.com/moorara))

## [v0.3.3](https://github.com/moorara/observer/tree/v0.3.3) (2020-09-23)

[Full Changelog](https://github.com/moorara/observer/compare/v0.3.2...v0.3.3)

**Closed issues:**

- Add RequestID to response header/metadata [\#62](https://github.com/moorara/observer/issues/62)

**Merged pull requests:**

- Add request id and client name to response headers [\#63](https://github.com/moorara/observer/pull/63) ([moorara](https://github.com/moorara))

## [v0.3.2](https://github.com/moorara/observer/tree/v0.3.2) (2020-09-19)

[Full Changelog](https://github.com/moorara/observer/compare/v0.3.1...v0.3.2)

**Fixed bugs:**

- Create noop logger, meter, and tracer by default [\#60](https://github.com/moorara/observer/issues/60)

**Merged pull requests:**

- Fix the bug related to nil logger, meter, and tracer [\#61](https://github.com/moorara/observer/pull/61) ([moorara](https://github.com/moorara))

## [v0.3.1](https://github.com/moorara/observer/tree/v0.3.1) (2020-09-10)

[Full Changelog](https://github.com/moorara/observer/compare/v0.3.0...v0.3.1)

**Fixed bugs:**

- Fix the bug for reporting number of in-flight requests [\#58](https://github.com/moorara/observer/issues/58)

**Merged pull requests:**

- Recovering and reporting panics from handlers [\#59](https://github.com/moorara/observer/pull/59) ([moorara](https://github.com/moorara))
- Update module google.golang.org/grpc to v1.32.0 [\#57](https://github.com/moorara/observer/pull/57) ([renovate[bot]](https://github.com/apps/renovate))
- Update module go.uber.org/zap to v1.16.0 [\#56](https://github.com/moorara/observer/pull/56) ([renovate[bot]](https://github.com/apps/renovate))
- Update module google/uuid to v1.1.2 [\#55](https://github.com/moorara/observer/pull/55) ([renovate[bot]](https://github.com/apps/renovate))

## [v0.3.0](https://github.com/moorara/observer/tree/v0.3.0) (2020-08-30)

[Full Changelog](https://github.com/moorara/observer/compare/v0.2.3...v0.3.0)

**Closed issues:**

- Provide an option to use OTEL collector [\#52](https://github.com/moorara/observer/issues/52)
- Consider renaming options [\#33](https://github.com/moorara/observer/issues/33)

**Merged pull requests:**

- Otel collector [\#54](https://github.com/moorara/observer/pull/54) ([moorara](https://github.com/moorara))
- Upgrade otel [\#53](https://github.com/moorara/observer/pull/53) ([moorara](https://github.com/moorara))
- Update module google.golang.org/grpc to v1.31.1 [\#51](https://github.com/moorara/observer/pull/51) ([renovate[bot]](https://github.com/apps/renovate))
- Housekeeping [\#46](https://github.com/moorara/observer/pull/46) ([moorara](https://github.com/moorara))

## [v0.2.3](https://github.com/moorara/observer/tree/v0.2.3) (2020-07-31)

[Full Changelog](https://github.com/moorara/observer/compare/v0.2.2...v0.2.3)

**Merged pull requests:**

- Update repo [\#45](https://github.com/moorara/observer/pull/45) ([moorara](https://github.com/moorara))
- Update module google.golang.org/grpc to v1.31.0 [\#44](https://github.com/moorara/observer/pull/44) ([renovate[bot]](https://github.com/apps/renovate))
- Update module go.opentelemetry.io/otel to v0.9.0 [\#42](https://github.com/moorara/observer/pull/42) ([renovate[bot]](https://github.com/apps/renovate))
- Update module go.opentelemetry.io/otel to v0.8.0 [\#41](https://github.com/moorara/observer/pull/41) ([renovate[bot]](https://github.com/apps/renovate))

## [v0.2.2](https://github.com/moorara/observer/tree/v0.2.2) (2020-07-03)

[Full Changelog](https://github.com/moorara/observer/compare/v0.2.1...v0.2.2)

**Merged pull requests:**

- Update OpenTelemetry module to v0.7.0 [\#40](https://github.com/moorara/observer/pull/40) ([moorara](https://github.com/moorara))
- Update module google.golang.org/protobuf to v1.25.0 [\#38](https://github.com/moorara/observer/pull/38) ([renovate[bot]](https://github.com/apps/renovate))
- Update module prometheus/client\_golang to v1.7.1 [\#37](https://github.com/moorara/observer/pull/37) ([renovate[bot]](https://github.com/apps/renovate))
- Update module google.golang.org/grpc to v1.30.0 [\#36](https://github.com/moorara/observer/pull/36) ([renovate[bot]](https://github.com/apps/renovate))
- Add Go Doc README badges [\#35](https://github.com/moorara/observer/pull/35) ([moorara](https://github.com/moorara))

## [v0.2.1](https://github.com/moorara/observer/tree/v0.2.1) (2020-06-18)

[Full Changelog](https://github.com/moorara/observer/compare/v0.2.0...v0.2.1)

**Merged pull requests:**

- Update README.md [\#32](https://github.com/moorara/observer/pull/32) ([moorara](https://github.com/moorara))

## [v0.2.0](https://github.com/moorara/observer/tree/v0.2.0) (2020-06-18)

[Full Changelog](https://github.com/moorara/observer/compare/v0.1.0...v0.2.0)

**Implemented enhancements:**

- ohttp: report canonical url paths when reporting metrics [\#17](https://github.com/moorara/observer/issues/17)

**Fixed bugs:**

- Prometheus histogram metrics cannot be created [\#9](https://github.com/moorara/observer/issues/9)
- Prometheus metrics do not have labels [\#8](https://github.com/moorara/observer/issues/8)

**Closed issues:**

- Finish ogrpc package [\#27](https://github.com/moorara/observer/issues/27)
- Finish ohttp package [\#20](https://github.com/moorara/observer/issues/20)
- Add a request gauge instrument [\#16](https://github.com/moorara/observer/issues/16)

**Merged pull requests:**

- Add examples [\#31](https://github.com/moorara/observer/pull/31) ([moorara](https://github.com/moorara))

observer.go
// Package observer can be used for implementing observability using OpenTelemetry API.
// It aims to unify three pillars of observability in one single package that is easy-to-use and hard-to-misuse.
//
// An Observer encompasses a logger, a meter, and a tracer.
// It offers a single unified developer experience for enabling observability.
package observer

import (
	"context"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/hashicorp/go-multierror"
	"github.com/prometheus/client_golang/prometheus"
	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
	"google.golang.org/grpc/credentials"

	"go.opentelemetry.io/otel/api/global"
	"go.opentelemetry.io/otel/api/metric"
	"go.opentelemetry.io/otel/api/trace"
	"go.opentelemetry.io/otel/label"
	"go.opentelemetry.io/otel/sdk/metric/controller/pull"
	"go.opentelemetry.io/otel/sdk/metric/controller/push"
	"go.opentelemetry.io/otel/sdk/metric/processor/basic"
	"go.opentelemetry.io/otel/sdk/metric/selector/simple"
	"go.opentelemetry.io/otel/sdk/resource"
	"go.opentelemetry.io/otel/semconv"

	promexporter "go.opentelemetry.io/otel/exporters/metric/prometheus"
	otlpexporter "go.opentelemetry.io/otel/exporters/otlp"
	jaegerexporter "go.opentelemetry.io/otel/exporters/trace/jaeger"
	tracesdk "go.opentelemetry.io/otel/sdk/trace"
)

type closeFunc func() error

type endFunc func(context.Context) error

func endFuncFromFunc(fn func()) endFunc {
	return func(ctx context.Context) error {
		done := make(chan struct{}, 1)
		go func() {
			fn()
			done <- struct{}{}
		}()

		select {
		case <-done:
			return nil
		case <-ctx.Done():
			return ctx.Err()
		}
	}
}

func endFuncFromCloseFunc(close closeFunc) endFunc {
	return func(ctx context.Context) error {
		done := make(chan error, 1)
		go func() {
			done <- close()
		}()

		select {
		case <-done:
			return nil
		case <-ctx.Done():
			return ctx.Err()
		}
	}
}

// configs is used for configuring and creating an observer.
type configs struct {
	name        string
	version     string
	environment string
	region      string
	tags        map[string]string

	// Logger
	loggerEnabled bool
	loggerLevel   string

	// Prometheus
	prometheusEnabled bool

	// Jaeger
	jaegerEnabled           bool
	jaegerAgentEndpoint     string
	jaegerCollectorEndpoint string
	jaegerCollectorUserName string
	jaegerCollectorPassword string

	// OpenTelemetry
	opentelemetryEnabled              bool
	opentelemetryCollectorAddress     string
	opentelemetryCollectorCredentials credentials.TransportCredentials
}

func configsFromEnv() configs {
	c := configs{}

	c.name = os.Getenv("OBSERVER_NAME")
	c.version = os.Getenv("OBSERVER_VERSION")
	c.environment = os.Getenv("OBSERVER_ENVIRONMENT")
	c.region = os.Getenv("OBSERVER_REGION")

	c.tags = map[string]string{}
	for _, env := range os.Environ() {
		pair := strings.Split(env, "=")
		if strings.HasPrefix(pair[0], "OBSERVER_TAG_") {
			tag := strings.TrimPrefix(pair[0], "OBSERVER_TAG_")
			tag = strings.ToLower(tag)
			c.tags[tag] = pair[1]
		}
	}

	if val := os.Getenv("OBSERVER_LOGGER_ENABLED"); val != "" {
		c.loggerEnabled, _ = strconv.ParseBool(val)
	}

	c.loggerLevel = os.Getenv("OBSERVER_LOGGER_LEVEL")

	// Defaults
	if c.loggerLevel == "" {
		c.loggerLevel = "info"
	}

	if val := os.Getenv("OBSERVER_PROMETHEUS_ENABLED"); val != "" {
		c.prometheusEnabled, _ = strconv.ParseBool(val)
	}

	if val := os.Getenv("OBSERVER_JAEGER_ENABLED"); val != "" {
		c.jaegerEnabled, _ = strconv.ParseBool(val)
	}

	c.jaegerAgentEndpoint = os.Getenv("OBSERVER_JAEGER_AGENT_ENDPOINT")
	c.jaegerCollectorEndpoint = os.Getenv("OBSERVER_JAEGER_COLLECTOR_ENDPOINT")
	c.jaegerCollectorUserName = os.Getenv("OBSERVER_JAEGER_COLLECTOR_USERNAME")
	c.jaegerCollectorPassword = os.Getenv("OBSERVER_JAEGER_COLLECTOR_PASSWORD")

	// Defaults
	if c.jaegerAgentEndpoint == "" && c.jaegerCollectorEndpoint == "" {
		c.jaegerAgentEndpoint = "localhost:6831"
	}

	if val := os.Getenv("OBSERVER_OPENTELEMETRY_ENABLED"); val != "" {
		c.opentelemetryEnabled, _ = strconv.ParseBool(val)
	}

	c.opentelemetryCollectorAddress = os.Getenv("OBSERVER_OPENTELEMETRY_COLLECTOR_ADDRESS")

	// Defaults
	if c.opentelemetryCollectorAddress == "" {
		c.opentelemetryCollectorAddress = "localhost:55680"
	}

	return c
}

// Option is an optional configuration for an observer.
type Option func(*configs)

// WithMetadata is the option for specifying and reporting metadata.
// All arguments are optional.
func WithMetadata(name, version, environment, region string, tags map[string]string) Option {
	return func(c *configs) {
		c.name = name
		c.version = version
		c.environment = environment
		c.region = region
		c.tags = tags
	}
}

// WithLogger is the option for configuring the logger.
// The default log level is info.
func WithLogger(level string) Option {
	if level == "" {
		level = "info"
	}

	return func(c *configs) {
		c.loggerEnabled = true
		c.loggerLevel = level
	}
}

// WithPrometheus is the option for reporting metrics for Prometheus.
func WithPrometheus() Option {
	return func(c *configs) {
		c.prometheusEnabled = true
	}
}

// WithJaeger is the option for reporting traces to Jaeger.
// Only one of agentEndpoint or collectorEndpoint is required.
// collectorUserName and collectorPassword are optional.
// The default agent endpoint is localhost:6831.
func WithJaeger(agentEndpoint, collectorEndpoint, collectorUserName, collectorPassword string) Option {
	if agentEndpoint == "" && collectorEndpoint == "" {
		agentEndpoint = "localhost:6831"
	}

	return func(c *configs) {
		c.jaegerEnabled = true
		c.jaegerAgentEndpoint = agentEndpoint
		c.jaegerCollectorEndpoint = collectorEndpoint
		c.jaegerCollectorUserName = collectorUserName
		c.jaegerCollectorPassword = collectorPassword
	}
}

// WithOpenTelemetry is the option for reporting metrics and traces to OpenTelemetry Collector.
// collectorCredentials is optional. If not specified, the connection will be insecure.
// The default collector address is localhost:55680.
func WithOpenTelemetry(collectorAddress string, collectorCredentials credentials.TransportCredentials) Option {
	if collectorAddress == "" {
		collectorAddress = "localhost:55680"
	}

	return func(c *configs) {
		c.opentelemetryEnabled = true
		c.opentelemetryCollectorAddress = collectorAddress
		c.opentelemetryCollectorCredentials = collectorCredentials
	}
}

// Observer provides logging, metrics, and tracing capabilities for observability.
type Observer interface {
	// End flushes and closes the logger, meter, and tracer.
	End(context.Context) error

	// Name is returns the name of the observer.
	Name() string

	// Logger is used for accessing the logger.
	Logger() *zap.Logger

	// SetLogLevel changes the logging level.
	SetLogLevel(level zapcore.Level)

	// GetLogLevel returns the current logging level.
	GetLogLevel() zapcore.Level

	// Meter is used for accessing the meter.
	Meter() metric.Meter

	// Tracer is used for accessing the tracer.
	Tracer() trace.Tracer

	// ServeHTTP implements http.Handler interface. It serves the metrics endpoint for Prometheus metrics.
	ServeHTTP(w http.ResponseWriter, r *http.Request)
}

type observer struct {
	name         string
	logger       *zap.Logger
	loggerConfig *zap.Config
	meter        metric.Meter
	promHandler  http.Handler
	tracer       trace.Tracer
	endFuncs     []endFunc
}

// New creates a new observer.
// If setAsSingleton set to true, the created observer will be set as the singleton observer too.
// So, you can also access it using observer.Get() function.
func New(setAsSingleton bool, opts ...Option) Observer {
	c := configsFromEnv()
	for _, opt := range opts {
		opt(&c)
	}

	o := &observer{
		name:     c.name,
		endFuncs: []endFunc{},
	}

	if c.loggerEnabled {
		o.logger, o.loggerConfig = initLogger(c)
		o.endFuncs = append(o.endFuncs,
			endFuncFromCloseFunc(o.logger.Sync),
		)
	}

	if c.prometheusEnabled {
		o.meter, o.promHandler = initPrometheus(c)
	}

	if c.jaegerEnabled {
		var tracerEnd endFunc
		o.tracer, tracerEnd = initJaeger(c)
		o.endFuncs = append(o.endFuncs, tracerEnd)
	}

	if c.opentelemetryEnabled {
		var meterEnd, tracerEnd endFunc
		o.meter, o.tracer, meterEnd, tracerEnd = initOpenTelemetry(c)
		o.endFuncs = append(o.endFuncs, meterEnd, tracerEnd)
	}

	// Create noop logger, meter, and/or tracer if they are not created so far

	if o.logger == nil {
		o.logger = zap.NewNop()
	}

	if o.loggerConfig == nil {
		o.loggerConfig = &zap.Config{}
	}

	if o.meter == (metric.Meter{}) {
		o.meter = new(metric.NoopMeterProvider).Meter("")
	}

	if o.promHandler == nil {
		o.promHandler = http.NotFoundHandler()
	}

	if o.tracer == nil {
		o.tracer = trace.NoopTracerProvider().Tracer("")
	}

	// Assign the new observer to the singleton observer
	if setAsSingleton {
		singleton = o
	}

	return o
}

func initLogger(c configs) (*zap.Logger, *zap.Config) {
	config := zap.Config{
		Level:       zap.NewAtomicLevelAt(zapcore.InfoLevel),
		Development: false,
		Sampling:    nil,
		Encoding:    "json",
		EncoderConfig: zapcore.EncoderConfig{
			TimeKey:        "timestamp",
			LevelKey:       "level",
			NameKey:        "logger",
			MessageKey:     "message",
			CallerKey:      "caller",
			StacktraceKey:  "stacktrace",
			LineEnding:     zapcore.DefaultLineEnding,
			EncodeLevel:    zapcore.LowercaseLevelEncoder,
			EncodeTime:     zapcore.RFC3339NanoTimeEncoder,
			EncodeCaller:   zapcore.ShortCallerEncoder,
			EncodeDuration: zapcore.SecondsDurationEncoder,
		},
		OutputPaths:      []string{"stdout"},
		ErrorOutputPaths: []string{"stdout"},
		InitialFields:    make(map[string]interface{}),
	}

	if c.name != "" {
		config.InitialFields["logger"] = c.name
	}

	if c.version != "" {
		config.InitialFields["version"] = c.version
	}

	if c.environment != "" {
		config.InitialFields["environment"] = c.environment
	}

	if c.region != "" {
		config.InitialFields["region"] = c.region
	}

	for k, v := range c.tags {
		config.InitialFields[k] = v
	}

	switch strings.ToLower(c.loggerLevel) {
	case "debug":
		config.Level = zap.NewAtomicLevelAt(zapcore.DebugLevel)
	case "info":
		config.Level = zap.NewAtomicLevelAt(zapcore.InfoLevel)
	case "warn":
		config.Level = zap.NewAtomicLevelAt(zapcore.WarnLevel)
	case "error":
		config.Level = zap.NewAtomicLevelAt(zapcore.ErrorLevel)
	case "none":
		fallthrough
	default:
		config.Level = zap.NewAtomicLevelAt(zapcore.Level(99))
	}

	logger, _ := config.Build(
		zap.AddCaller(),
		zap.AddCallerSkip(0),
	)

	return logger, &config
}

func initPrometheus(c configs) (metric.Meter, http.Handler) {
	// Create a new Prometheus registry
	registry := prometheus.NewRegistry()
	registry.MustRegister(prometheus.NewGoCollector())
	registry.MustRegister(prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}))

	config := promexporter.Config{
		Registerer: registry,
		Gatherer:   registry,
		// DefaultSummaryQuantiles:    []float64{},
		// DefaultHistogramBoundaries: []float64{},
	}

	pullOpts := []pull.Option{}
	exporter, err := promexporter.NewExportPipeline(config, pullOpts...)
	if err != nil {
		panic(err)
	}

	global.SetMeterProvider(exporter.MeterProvider())
	meter := exporter.MeterProvider().Meter(c.name)

	return meter, exporter
}

func initJaeger(c configs) (trace.Tracer, endFunc) {
	var endpointOpt jaegerexporter.EndpointOption
	switch {
	case c.jaegerAgentEndpoint != "":
		endpointOpt = jaegerexporter.WithAgentEndpoint(c.jaegerAgentEndpoint)
	case c.jaegerCollectorEndpoint != "":
		endpointOpt = jaegerexporter.WithCollectorEndpoint(
			c.jaegerCollectorEndpoint,
			jaegerexporter.WithUsername(c.jaegerCollectorUserName),
			jaegerexporter.WithPassword(c.jaegerCollectorPassword),
		)
	}

	tags := []label.KeyValue{}
	for k, v := range c.tags {
		tags = append(tags, label.String(k, v))
	}

	processOpt := jaegerexporter.WithProcess(
		jaegerexporter.Process{
			ServiceName: c.name,
			Tags:        tags,
		},
	)

	sdkOpt := jaegerexporter.WithSDK(
		&tracesdk.Config{
			DefaultSampler: tracesdk.AlwaysSample(),
		},
	)

	provider, flush, err := jaegerexporter.NewExportPipeline(endpointOpt, processOpt, sdkOpt)
	if err != nil {
		panic(err)
	}

	global.SetTracerProvider(provider)
	tracer := global.TracerProvider().Tracer(c.name)

	return tracer, endFuncFromFunc(flush)
}

func initOpenTelemetry(c configs) (metric.Meter, trace.Tracer, endFunc, endFunc) {
	// Exporter

	expOpts := []otlpexporter.ExporterOption{
		otlpexporter.WithAddress(c.opentelemetryCollectorAddress),
	}
	if c.opentelemetryCollectorCredentials == nil {
		expOpts = append(expOpts, otlpexporter.WithInsecure())
	} else {
		expOpts = append(expOpts, otlpexporter.WithTLSCredentials(c.opentelemetryCollectorCredentials))
	}

	exporter, err := otlpexporter.NewExporter(expOpts...)
	if err != nil {
		panic(err)
	}

	// Trace Provider

	tpOpts := []tracesdk.TracerProviderOption{
		tracesdk.WithBatcher(exporter),
		tracesdk.WithConfig(tracesdk.Config{
			DefaultSampler: tracesdk.AlwaysSample(),
		}),
		tracesdk.WithResource(resource.New(
			semconv.ServiceNameKey.String(c.name),
		)),
	}

	traceProvider := tracesdk.NewTracerProvider(tpOpts...)

	// Meter Provider

	aggregator := simple.NewWithExactDistribution()
	checkpointer := basic.New(aggregator, exporter)
	pushOpts := []push.Option{
		push.WithPeriod(2 * time.Second),
	}

	pusher := push.New(checkpointer, exporter, pushOpts...)

	// Set global providers
	global.SetTracerProvider(traceProvider)
	global.SetMeterProvider(pusher.MeterProvider())
	pusher.Start()

	meter := global.MeterProvider().Meter(c.name)
	tracer := global.TracerProvider().Tracer(c.name)

	return meter, tracer, endFuncFromFunc(pusher.Stop), exporter.Shutdown
}

func (o *observer) End(ctx context.Context) error {
	var err error
	for _, endFunc := range o.endFuncs {
		if e := endFunc(ctx); e != nil {
			err = multierror.Append(err, e)
		}
	}

	return err
}

func (o *observer) Name() string {
	return o.name
}

func (o *observer) Logger() *zap.Logger {
	return o.logger
}

func (o *observer) SetLogLevel(level zapcore.Level) {
	o.loggerConfig.Level.SetLevel(level)
}

func (o *observer) GetLogLevel() zapcore.Level {
	return o.loggerConfig.Level.Level()
}

func (o *observer) Meter() metric.Meter {
	return o.meter
}

func (o *observer) Tracer() trace.Tracer {
	return o.tracer
}

func (o *observer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	if o.promHandler != nil {
		o.promHandler.ServeHTTP(w, r)
	}
}

var singleton *observer

// Initialize the singleton observer with a no-op observer.
// init function will be only called once in runtime regardless of how many times the package is imported.
func init() {
	singleton = &observer{
		logger:       zap.NewNop(),
		loggerConfig: &zap.Config{},
		meter:        new(metric.NoopMeterProvider).Meter(""),
		promHandler:  http.NotFoundHandler(),
		tracer:       trace.NoopTracerProvider().Tracer(""),
	}
}

// Get returns the singleton Observer.
func Get() Observer {
	return singleton
}
